File: src/bin/77_combinations.py
#
# @lc app=leetcode id=77 lang=python3
#
# [77] Combinations
#


# @lc code=start
import copy


class Solution:
    def combine(self, n: int, k: int) -> list[list[int]]:
        answer: list[list[int]] = []
        combination = list(range(1, k + 1))
        while True:
            answer.append(copy.copy(combination))
            for i in reversed(range(k)):
                if combination[i] != n - k + i + 1:
                    break
            else:
                break
            combination[i] += 1
            for j in range(i + 1, k):
                combination[j] = combination[j - 1] + 1

        return answer


def main():
    n, k = 4, 2
    print(Solution().combine(n, k))


# @lc code=end

if __name__ == "__main__":
    main()

File: src/bin/77_combinations.rs
/*
 * @lc app=leetcode id=77 lang=rust
 *
 * [77] Combinations
 */

// @lc code=start
impl Solution {
    pub fn combine(n: i32, k: i32) -> Vec<Vec<i32>> {
        let mut answer = vec![];
        let mut combination: Vec<i32> = (1..=k).collect();
        loop {
            answer.push(combination.clone());
            let Some((pivot_index, _)) = combination
                .iter()
                .enumerate()
                .rev()
                .find(|&(i, &elem)| elem != n - k + i as i32 + 1)
            else {
                break;
            };
            combination[pivot_index] += 1;
            for j in pivot_index + 1..k as usize {
                combination[j] = combination[j - 1] + 1;
            }
        }
        answer
    }
}

fn _main() {
    let n = 4;
    let k = 2;
    println!("{:?}", Solution::combine(n, k))
}

// @lc code=end

struct Solution;

fn main() {
    _main()
}
